mcollina/autocannon
# feat: support HAR files as input

#### What's in there?

- [x] support for [HAR](https://w3c.github.io/web-performance/specs/HAR/Overview.html) files as request input
- [x] closes #185

This PR allow CLI users to speciffy `--har`  options with the path to an `.har`  file.
Requests from the har will be fired against the specify server.

Please note that Autocannon will ignore HAR requests that would not match the domain of your URL.
This is because each httpClient handles the connection to a single domain. It is not possible for a client to switch to a different one, based on current request. 

#### How to test?

You can use any of the test fixtures:
```shell
node autocannon --har ./test/fixtures/httpbin-get.har -d 3 https://httpbin.org
``` 

You can also record an .har from your favorite browser's dev tools. 
For example, with you devtools opened, go the [node.js reference](https://nodejs.org/api/url.html#url_legacy_url_api). A couple of requests will appear on network tab. 
Export them as an HAR file, and try playing it again:

```shell
node autocannon --har test.har -d 3 -j https://nodejs.org
``` 

Please note that some request may "fail" with "3xx" statuses.

Finally, it is possible to specify `har`  option when using the programmatic API, as demonstrated in the `tests/run.test.js` file.

#### Notes to reviewers

- HAR files are plain JSON files, and there is no need for a specific parser, in particular because we're using a tiny part of the file (we ignore pages, responses, timing...)

- `parseHAR.js`  file is exposing functions to read and parse files in a synchronous or asynchronous fashion. However, it made little sense to use the _asynchronous_ flavor: HAR parsing occurs only once, before starting benchmark, and that part of Autocannon code if synchronous.
Changing it to asynchronous would be a major change for little-to-no benefit (in my opinion).

- there could be more checks on the HAR content (I assumed that requests will contain the proper keys), and I'm happy to add more if you think it worth it.
